File: WithingsAuthApp/app/src/main/java/com/example/withingsauthapp/MainActivity.java
package com.example.withingsauthapp;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    private static final String CLIENT_ID = "82108352dc5fb2557c0686664160c4dcd047357dd0bb9110d3383bd9f0f1e3de\n";
    private static final String REDIRECT_URI = "https://oauth.pstmn.io/v1/callback"; // URL de redirection
    private static final String AUTH_URL = "https://account.withings.com/oauth2_user/authorize2?response_type=code&client_id=82108352dc5fb2557c0686664160c4dcd047357dd0bb9110d3383bd9f0f1e3de&redirect_uri=https%3A%2F%2Foauth.pstmn.io%2Fv1%2Fcallback&state=8aBKns8AReUfdxC&scope=user.metrics%2Cuser.activity" +
            "?response_type=code" +
            "&client_id=" + CLIENT_ID +
            "&redirect_uri=" + REDIRECT_URI +
            "&scope=user.info,user.metrics" +  // Ajouter les scopes nécessaires
            "&state=random_string";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button authButton = findViewById(R.id.authButton);
        authButton.setOnClickListener(view -> startOAuth());
    }

    private void startOAuth() {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(AUTH_URL));
        startActivity(browserIntent);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        Uri uri = intent.getData();

        if (uri != null && uri.toString().startsWith("withingsauthapp://callback")) {
            String authCode = uri.getQueryParameter("code");
            if (authCode != null) {
                Log.d("OAuth", "Code reçu : " + authCode);
                getAccessToken(authCode);
            } else {
                Log.e("OAuth", "Erreur lors de la récupération du code");
            }
        }
    }

    private void getAccessToken(String authCode) {
        // Ici, tu dois envoyer une requête POST à Withings pour échanger le code contre un access_token
        Log.d("OAuth", "Échange du code contre un access_token...");
    }
}

File: Balance2/app/src/main/java/com/example/balance2/MainActivity.kt
package com.example.balance2

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import okhttp3.*
import org.json.JSONObject
import java.io.IOException

class MainActivity : AppCompatActivity() {

    private val CLIENT_ID = "TON_CLIENT_ID"
    private val CLIENT_SECRET = "TON_CLIENT_SECRET"
    private val REDIRECT_URI = "tonapp://callback"  // Assure-toi de l’enregistrer dans Withings Developer
    private val AUTH_URL = "https://account.withings.com/oauth2_user/authorize2" +
            "?response_type=code" +
            "&client_id=$CLIENT_ID" +
            "&redirect_uri=$REDIRECT_URI" +
            "&scope=user.metrics" +
            "&state=12345"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Ouvre la page d'authentification de Withings
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(AUTH_URL))
        startActivity(intent)
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        val uri = intent?.data
        if (uri != null && uri.toString().startsWith(REDIRECT_URI)) {
            val authCode = uri.getQueryParameter("code")
            if (authCode != null) {
                Log.d("OAuth", "Code reçu : $authCode")
                getAccessToken(authCode)
            } else {
                Log.e("OAuth", "Erreur lors de la récupération du code d'authentification")
            }
        }
    }

    private fun getAccessToken(authCode: String) {
        val url = "https://wbsapi.withings.net/v2/oauth2"

        // Configuration sécurisée du client OkHttp
        val client = OkHttpClient.Builder()
            .build()

        val body = FormBody.Builder()
            .add("action", "requesttoken")
            .add("client_id", CLIENT_ID)
            .add("client_secret", CLIENT_SECRET)
            .add("code", authCode)
            .add("grant_type", "authorization_code")
            .add("redirect_uri", REDIRECT_URI)
            .build()

        val request = Request.Builder()
            .url(url)  // 🚀 Toujours en HTTPS
            .post(body)
            .build()

        client.newCall(request).enqueue(object : Callback
        {
            override fun onFailure(call: Call, e: IOException)
            {
                Log.e("OAuth", "Échec de la requête HTTPS : ${e.message}")
            }

            override fun onResponse(call: Call, response: Response) {
                val responseData = response.body?.string()
                if (response.isSuccessful) {
                    val json = JSONObject(responseData ?: "")
                    val accessToken = json.getJSONObject("body").getString("access_token")
                    Log.d("OAuth", "Access Token sécurisé reçu : $accessToken")
                    // Stocker le token de manière sécurisée (ex: EncryptedSharedPreferences)
                } else {
                    Log.e("OAuth", "Erreur dans la réponse HTTPS : $responseData")
                }
            }
        })
    }
}